core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java

```diff
@@ -68,19 +69,71 @@
 
   protected final String logPrefix;
   private final TypeCodec<?>[] primitiveCodecs;
-  private final TypeCodec<?>[] userCodecs;
+  private final CopyOnWriteArrayList<TypeCodec<?>> userCodecs = new CopyOnWriteArrayList<>();
   private final IntMap<TypeCodec> primitiveCodecsByCode;
 
   protected CachingCodecRegistry(
-      @NonNull String logPrefix,
-      @NonNull TypeCodec<?>[] primitiveCodecs,
-      @NonNull TypeCodec<?>[] userCodecs) {
+      @NonNull String logPrefix, @NonNull TypeCodec<?>[] primitiveCodecs) {
     this.logPrefix = logPrefix;
     this.primitiveCodecs = primitiveCodecs;
-    this.userCodecs = userCodecs;
     this.primitiveCodecsByCode = sortByProtocolCode(primitiveCodecs);
   }
 
+  /**
+   * @deprecated this constructor calls an overridable method ({@link #register(TypeCodec[])}),
+   *     which is a bad practice. The recommended alternative is to use {@link
+   *     #CachingCodecRegistry(String, TypeCodec[])}, then add the codecs with one of the {@link
+   *     #register} methods.
+   */
+  @Deprecated
+  protected CachingCodecRegistry(
+      @NonNull String logPrefix,
+      @NonNull TypeCodec<?>[] primitiveCodecs,
+      @NonNull TypeCodec<?>[] userCodecs) {
+    this(logPrefix, primitiveCodecs);
+    register(userCodecs);
+  }
```

Comment:
Deprecating this constructor is the best way I've found to address the fact that it calls an overridable method. 3.x didn't have a way to pass user codecs to the constructor so I don't think it's a big deal to drop that.

Reply:
It should be possible to avoid the deprecation of this constructor:

```
 protected CachingCodecRegistry(
    @NonNull String logPrefix,
    @NonNull TypeCodec<?>[] primitiveCodecs,
    @NonNull TypeCodec<?>[] userCodecs) {
  this.logPrefix = logPrefix;
  this.primitiveCodecs = primitiveCodecs;
  this.primitiveCodecsByCode = sortByProtocolCode(primitiveCodecs);
  for (TypeCodec<?> userCodec : userCodecs) {
    registerInternal(userCodec);
  }
}

@Override
public void register(TypeCodec<?> newCodec) {
  registerInternal(newCodec);
}

private void registerInternal(TypeCodec<?> newCodec) {
  ...
}
```

Reply:
I thought of that, but then if you have a subclass that needs to do some additional processing when you register a codec, it can only do so in the overridden method, not change how the constructor handles registration.